src/main/java/org/apache/commons/lang3/reflect/TypeUtils.java

```diff
@@ -1811,9 +1823,14 @@ private static String parameterizedTypeToString(final ParameterizedType p) {
         return buf.toString();
     }
 
-    private static void appendRecursiveTypes(final StringBuilder buf, final int[] recursiveTypeIndexes, final Type[] argumentTypes) {
-        for (int i = 0; i < recursiveTypeIndexes.length; i++) {
-            appendAllTo(buf.append('<'), ", ", argumentTypes[i].toString()).append('>');
+    @SuppressWarnings({"compound.assignment.type.incompatible","array.access.unsafe.high","assignment.type.incompatible"})/*
+    for (@IndexOrHigh("argumentTypes") int i = 0; i < recursiveTypeIndexes.length; i++) // i <= argumentTypes.length
+    appendAllTo(buf.append('<'), ", ", argumentTypes[i].toString()).append('>'); // i is a valid index of argumentTypes
+    indexesToRemove = ArrayUtils.add(indexesToRemove, i); // ArrayUtils.add(indexesToRemove, i) at the end of the loop returns an array with maximum possible length of p.getActualTypeArguments().length 
+    */
+    private static void appendRecursiveTypes(final StringBuilder buf, final int @LTEqLengthOf("#3") [] recursiveTypeIndexes, final Type[] argumentTypes) {
+        for (@IndexOrHigh("argumentTypes") int i = 0; i < recursiveTypeIndexes.length; i++) { // i <= argumentTypes.length
+            appendAllTo(buf.append('<'), ", ", argumentTypes[i].toString()).append('>'); // i is a valid index of argumentTypes 
```

Comment:
It looks like `argumentTypes` is the same length as `recursiveTypeIndexes`, but that's not expressed in the annotations.